### PR TITLE
Remove non position-independent pseudo instructions (ldr rx,=label)

### PR DESCRIPTION
**Pull request by h6ah4i**
This commit fixes the issue #83.

Approach 1: (for NE10_fft_float32.neon.s, NE10_fft_float32.neonv8.S, NE10_fft_int16.neon.s)
- Use "adr" pseudo instruction

Approach 2: (for NE10_fir.neon.s, NE10_iir.neon.s)
- Use GOT based access

### CODE REVIEW

**Comment by ashi08104** [comment]
Hi, there,
Thanks for your contribution and sorry for taking so long to respone this issue.

Ne10 also support iOS, does your modification also work for iOS? However, the situation is a little hard currently: Ne10's iOS support is broken now, and we have not much bandwitdh to fix it. We can't easily test your modification for iOS platform. So it is also OK that if you are not sure this modification can work on iOS.

**Comment by h6ah4i** [comment]
Hi. Unfortunately, I'm not sure whether my modification works well on iOS. Because I don't have any iOS devices :(

**Comment by zhongweiy** [comment]
It's OK. I will look at this change when the legal process is done.

**Comment by sbergen** [comment]
Please consider pushing forward with this pull request, as it is **required** if you want to use the latest Android SDKs!

**Comment by Phillip-Wang** [comment]
I think we should make some progress here. @zhongweiy , would you have a look?

**Comment by marforic** [comment]
This patch allows Ne10 to work against Android API v23 (Latest from Google, Android 6). I'd love to see this merged so I don't have to manually patch and re-compile my version of the library.

File: modules/dsp/NE10_fir.neon.s [hunk 1]
```diff
@@ -532,10 +541,19 @@ dInp2_1          .dn   D31.F32
...
                     @//outBlockSize = blockSize / S->M
+#ifdef __PIC__
+                    @/* position-independent access of LDR pMask,=ne10_divLookUpTable */
+                    LDR         pTemp,.L_PIC1_GOT_OFFSET
+                    LDR         pMask,.L_GOT_ne10_divLookUpTable
+.L_PIC1:
+                    ADD         pTemp,pTemp,pc
+                    LDR         pMask,[pTemp,pMask]
+#else
                     LDR         pTemp,=ne10_divLookUpTable
```
Comment (marforic) [inline modules/dsp/NE10_fir.neon.s]: Should this be changed to?

``LDR         pMask,=ne10_divLookUpTable``

Comment (h6ah4i) [inline modules/dsp/NE10_fir.neon.s]: I think so too. But I have not changed this line though... :confused:

Comment (Brahmasmi) [inline modules/dsp/NE10_fir.neon.s]: @h6ah4i san, should this line be changed or not? Unfortunately, I possess no knowledge of ARM assembly/intrinsics, so I cannot determine if the change suggested by @marforic should be made or not.

**Comment by Brahmasmi** [comment]
@Phillip-Wang @zhongweiy Would it be possible for you to please share a tentative timeline for resolving this?

As mentioned previously, Android 23 and above do not support text relocations. From https://developer.android.com/about/versions/marshmallow/android-6.0-changes.html#behavior-runtime (emphasis added)-
<pre><code>
On previous versions of Android, if your app requested the system to load
a shared library with text relocations, the system displayed a warning but
still allowed the library to be loaded. Beginning in this release, the system
<b>rejects</b> this library if your app's target SDK version is 23 or higher.
</code></pre>
As such, it is currently not possible to use this excellent library.

File: modules/dsp/NE10_fir.neon.s (hunk 1 above)
Comment (h6ah4i) [inline modules/dsp/NE10_fir.neon.s]: @Brahmasmi Sorry, I cannot answer without more investigation. At least, we should write a test code first.

**Comment by Brahmasmi** [comment]
@mdupuy Could you please help us get this merged?

The details are present in this pull request as well as the linked issue. This is important because Android 23 and above do not support text relocations, limiting the use of this excellent library.

File: modules/dsp/NE10_fir.neon.s (hunk 1 above)
Comment (joesavage) [inline modules/dsp/NE10_fir.neon.s]: I'm far from an expert on PIC, but I don't think that this line should be changed. [`LDR Rt, .label`](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0041c/Babbfdih.html) is an instruction to load a `pc` relative value within a certain range into a register, while `LDR Rt, =.label` is a pseudo-instruction that loads an absolute address into a register via a literal pool. In this case, the behaviour of the the former is what we want as `pTemp` and `pMask` are supposed to contain the words at those labels.

To outline more explicitly what's going on here, the `ADD` instruction sums the relative offset between the GOT and the end of the final `LDR` instruction (i.e. `_GLOBAL_OFFSET_TABLE_-(.L_PICX+4)`) with the actual address after the final `LDR` instruction at execution (i.e. `pc`, which is the absolute address of `(.L_PICX+4)` as we're in Thumb mode) to get the absolute address of the GOT. This is then used in combination with the offset of the data (e.g. `ne10_qMaskTable32`) within the lookup table to perform the position-independent load.

**Comment by joesavage** [comment]
I'm no PIC expert, but having had a quick look over the changes:

- The changes utilizing the GOT look good to me. From what I can tell this is a relatively standard pattern, and is often what gets generated by C compilers.
- The `ADR` changes generally seem fine, and may be beneficial even for non-position-independent code, however I am unsure if the addresses being loaded are always [aligned properly or within the 1020 byte range required](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0204f/Babcjaii.html). If the relevant blocks could be placed under an `.align` directive, and are assured to be in range, then these changes seem fine by me.
- Your changes to lines [555](https://github.com/projectNe10/Ne10/pull/84/files#diff-95c16ed3c25d44b3d3d3e794e4243b0aR555) and [556](https://github.com/projectNe10/Ne10/pull/84/files#diff-95c16ed3c25d44b3d3d3e794e4243b0aR556) are causing some FIR tests to fail on my machine with `__PIC__` left undefined. This is fairly straightfoward to fix, e.g. by changing [line 552](https://github.com/projectNe10/Ne10/pull/84/files#diff-95c16ed3c25d44b3d3d3e794e4243b0aR552. After making this change I'm finding that the results of the DSP module tests are unchanged from master in both the position dependent and independent versions of the code on my machine (woohoo!).

**Comment by joesavage** [comment]
Great! The only other thing that worries me now besides the possible `ADR` alignment and range issues is iOS support (I assume Android support has already been tested). I haven't been able to test if this compiles and works properly on iOS, but I have a suspicion that it might not work correctly. The reason being that [`ADR` is a pseudo-instruction](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0041c/Babcjaii.html), and as such may not be supported by clang's integrated assembler. [`gas2ios_convert.py`](https://github.com/projectNe10/Ne10/blob/master/tools/gas2ios_convert.py) currently deals with this for other pseudo-instructions (such as [`LDR Rt, =label`](http://infocenter.arm.com/help/topic/com.arm.doc.dui0041c/Babbfdih.html)), but does not appear to currently perform any translation for `ADR`.

Either way, we ought to try and get somebody to test this on iOS at some stage or another before merging. Hopefully sooner rather than later, because this is an important patch! (Thanks for contributing it!)

**Comment by h6ah4i** [comment]
@joesavage Hi. Thank you for reviewing the pull request. I confirmed the line 552 bug was my fault and it has been fixed by the commit 61d5bbb3fc2ad7c176c7f47140cde23acac4aca4. Also the base master branch was too old (2 years ago...!), so I merged the latest master branch.

I am not sure that `ADR` pseudo instruction is compatible with iOS toolchains, but at least I have succeeded to build binary by the following commands. However, I do not have any iOS devices for now 😞 

```bash
$ mkdir build && cd build

$ cmake -DCMAKE_TOOLCHAIN_FILE=../ios/ios_config.cmake -DMIN_IOS_VER=7.0  ..
-- The C compiler identification is AppleClang 8.0.0.8000038
-- The CXX compiler identification is AppleClang 8.0.0.8000038
-- The ASM compiler identification is unknown
-- Found assembler: /usr/bin/TRUE
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - failed
-- Detecting C compile features
-- Detecting C compile features - failed
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - failed
-- Detecting CXX compile features
-- Detecting CXX compile features - failed
-- Warning: Did not find file Compiler/-ASM
-- Target architecture : armv7
-- Building type: RELEASE
-- CFLAGS:  -O2 -DNDEBUG -miphoneos-version-min=;7.0 -mfpu=vfp3 -arch armv7 -arch armv7s
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/hasegawa/git/Ne10/build

$ make -j4
```

**Comment by joesavage** [comment]
No problem, thanks for contributing this important patch! I'm not sure whether this project generally prefers merge commits or rebasing, but the git history looks pretty messy anyway so I guess it doesn't really matter too much as long as there aren't any conflicts. Do you have any thoughts on the potential `ADR` range/alignment issues I was talking about?

Also, it's great news that you got everything to compile properly with clang. I guess maybe it does support `ADR` after all. I've tried testing with my Macbook/iPhone setup, and everything appears to work properly (compiles fine and runs [presumably correctly, it just spits out some performance numbers]). I have no idea if I'm testing a configuration that actually uses the files that have been modified in this patch though, as I can't see any mention of the optimised assembly files in the output from `make`.

**Comment by h6ah4i** [comment]
@joesavage 

> project generally prefers merge commits or rebasing, but the git history looks pretty messy anyway

I think this is not a matter anymore because GitHub can rebase/squash PR when merging. (BTW, I chose merge commit in this case because we've talked about specific line number.)

image

> so I guess it doesn't really matter too much as long as there aren't any conflicts. 

I agree, but latest master branch code was required to pass iOS build. I will push old commit without merging, if project manager asks me.


> Do you have any thoughts on the potential ADR range/alignment issues I was talking about?

I think ADR range limit will not cause a problem in most of case.

I made a small experiment to test assembler's behavior. Putting many `nop` instructions (= 2 bytes / instruction) right before invoking the `BUTTERFLY8X2_WITHOUT_TWIDDLES` macro. 

Here is the result:
- 512 `nop`(1k bytes) : **no problem**
- 1k `nop` (2k bytes): **no problem**
  - GAS automatically translate into [Thumb-2 (32 bit) instruction](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0204f/Babcjaii.html), it allows PC relative +/- 4k bytes accessing.
- 2k `nop` (4k bytes): **assemble failed**

  > .../Ne10/modules/dsp/NE10_fft_float32.neon.s: Assembler messages:
  > .../Ne10/modules/dsp/NE10_fft_float32.neon.s:786: Error: invalid constant (1484) after fixup

However, I think alignment should be explicitly specified. Because Thumb-2's ADR pseudo instruction also allows byte or halfword-aligned address. This can lead unexpected paddings between the `.L_TW_81` label and two float values.

> 32-bit Thumb-2
> ±4095 bytes to a byte, halfword, or word-aligned address.

**Comment by joesavage** [comment]
I see, thanks for the detailed response. Given the results of your assembler tests, I think this patch should be good to merge after the proper alignment directives are added. (It's only taken two years!)

**Pull request merged**